modules/viewabilityScoreGeneration.js

```diff
@@ -3,110 +3,180 @@ import adapterManager from '../src/adapterManager.js';
+        if (Object.keys(adSizes).length) bidViewabilityFields.adSizes = adSizes;
+        if (adUnit) bidViewabilityFields.adUnit = removeKeys(deepClone(adUnit));
+        if (Object.keys(bidViewabilityFields).length) bid.bidViewability = bidViewabilityFields;
       });
     });
   }
 
   fn(bidderRequests);
-}
+};
 
-export const gptSlotRenderEndedHandler = (adSlotElementId, setToLocalStorageCb) => {
-  if (vsgObj) {
-    if (vsgObj[adSlotElementId]) {
-      if (vsgObj[adSlotElementId].lastViewed) delete vsgObj[adSlotElementId].lastViewed;
+const removeKeys = obj => {
+  // Deleteing this field as it is only required to calculate totalViewtime and no need to send it to translator.
+  delete obj.lastViewStarted;
+  // Deleteing totalTimeView incase value is less than 1 sec.
+  if (obj.totalViewTime == 0) {
+    delete obj.totalViewTime;
+  }
+  // Deleting slot field as it is only required to pass correct size values in corresponding impressions.
+  delete obj.slot;
+  return obj;
+};
 
-      vsgObj[adSlotElementId].rendered = vsgObj[adSlotElementId].rendered + 1;
-      vsgObj[adSlotElementId].updatedAt = Date.now();
-    } else {
-      vsgObj[adSlotElementId] = {
-        rendered: 1,
-        viewed: 0,
-        createdAt: Date.now()
-      }
-    }
+// once the TOTAL_VIEW_TIME_LIMIT for totalViewTime is reached, divide totalViewTime, rendered & viewed all by the same factor of "x" in order to preserve the same averages but not let counts in localstorage get too high
+const reduceAndPreserveCounts = (key, lsObj = vsgObj) => {
+  const divideBy = 2;
+  lsObj[key].totalViewTime = Math.round(lsObj[key].totalViewTime / divideBy);
+  lsObj[key].rendered = Math.round(lsObj[key].rendered / divideBy);
+  lsObj[key].viewed = Math.round(lsObj[key].viewed / divideBy);
+};
+
+export const updateTotalViewTime = (diff, currentTime, lastViewStarted, key, lsObj = vsgObj) => {
+  diff = currentTime - lastViewStarted;
+  const newValue = Math.round((lsObj[key].totalViewTime || 0) + diff / 1000);
+
+  if (newValue >= TOTAL_VIEW_TIME_LIMIT) {
+    reduceAndPreserveCounts(key, lsObj);
   } else {
-    vsgObj = {
-      [adSlotElementId]: {
-        rendered: 1,
-        viewed: 0,
-        createdAt: Date.now()
-      }
-    }
+    lsObj[key].totalViewTime = newValue;
   }
-
-  setToLocalStorageCb('viewability-data', vsgObj);
 };
 
-export const gptImpressionViewableHandler = (adSlotElementId, setToLocalStorageCb) => {
-  if (vsgObj) {
-    if (vsgObj[adSlotElementId]) {
-      vsgObj[adSlotElementId].viewed = vsgObj[adSlotElementId].viewed + 1;
-      vsgObj[adSlotElementId].updatedAt = Date.now();
+// function to return default values for rendered, viewed, slot and createdAt
+// slot is required for getting correct values from local storage
+const defaultInit = (keyArr, index) => {
+	return {
+		rendered: 1,
+        viewed: 0,
+		slot: index == ADSLOTSIZE_INDEX ? [keyArr[ADUNIT_INDEX]] : undefined,
```

Comment:
noticed that the `slot` field is new.  what is this for?  sounds like it is to help keep track of ad sizes per impression?